README.md

# catplot: Capable And Tity Plot <img src="man/figures/logo.png" align="right" alt="" width="120" />

[![R-CMD-check](https://github.com/zerostwo/catplot/workflows/R-CMD-check/badge.svg)](https://github.com/zerostwo/catplot)
[![Codecov test
coverage](https://codecov.io/gh/zerostwo/catplot/branch/main/graph/badge.svg)](https://app.codecov.io/gh/zerostwo/catplot?branch=main)
[![lifecycle](https://img.shields.io/badge/lifecycle-Experimental-important.svg)](https://lifecycle.r-lib.org/articles/stages.html)

<!-- README.md is generated from README.Rmd. Please edit that file -->

## 🤪 Overview

`catplot` is a capable and tity data visualization tool and maintained
by [Songqi Duan](https://songqi.online).

## 📦 Installation

Install the latest version from
[GitHub](https://github.com/zerostwo/catplot) as follow:

``` r
if (!require(devtools)) install.packages("devtools")
devtools::install_github("zerostwo/catplot")
```

## 🕹️ Usage

``` r
library(catplot)
library(ggplot2)

data("iris")
p <- ggplot(
  data = iris,
  aes(
    x = Species,
    y = Sepal.Length
  )
) +
  geom_boxplot() +
  theme_cat(
    aspect_ratio = 1,
    show_panel_grid = "both",
    show_title = "y"
  )
p
```

<img src="man/figures/README-iris-1.png" width="100%" />

## 🧩 Code of Conduct

Please note that the catplot project is released with a [Contributor
Code of
Conduct](https://contributor-covenant.org/version/2/1/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# catplot 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


R/boxplot.R
#' Create a boxplot with customized options using ggplot2
#'
#' This function creates a boxplot using ggplot2 with options for modifying the appearance of the plot.
#'
#' @inheritParams cat_modify_plot
#' @param data A data frame containing the data to be plotted.
#' @param x The variable to be plotted on the x-axis.
#' @param y The variable to be plotted on the y-axis.
#' @param fill The variable to be used for filling the boxes. Defaults to NULL.
#' @param width The width of the boxes. Defaults to 0.6.
#' @param fatten The amount by which to fatten the boxes. Defaults to 1.
#' @param flip Logical indicating whether to flip the axes of the plot. Defaults to FALSE.
#' @param show_legend Logical indicating whether to show the legend. Defaults to FALSE.
#'
#' @return A ggplot2 object representing the boxplot with customized options.
#' @export
#'
#' @examples
#' library(catplot)
#' library(ggplot2)
#' library(dplyr)
#'
#' data("iris")
#' iris %>%
#'   mutate(group = if_else(log2(Petal.Length + 1) > 1.5,
#'     "high", "low"
#'   )) %>%
#'   cat_boxplot(
#'     x = Species,
#'     y = log2(Petal.Width + 1),
#'     aspect_ratio = 1,
#'     show_panel_grid = "both",
#'     fill = group,
#'     legend_title = "ee",
#'     legend_position = "top",
#'     x_text_angle = 45
#'   )

#' p <- iris |>
#'   cat_boxplot(
#'     x = Species, y = Sepal.Length,
#'     aspect_ratio = 2, frame = "open",
#'     x_text_angle = 45,
#'     comparisons = list(c("versicolor", "virginica"))
#'   ) + scale_fill_brewer(palette = "Paired")
#' p
cat_boxplot <- function(data,
                        x,
                        y,
                        fill = NULL,
                        width = 0.6,
                        fatten = 1,
                        show_legend = FALSE,
                        aspect_ratio = NULL,
                        flip = FALSE,
                        frame = "closed",
                        font_size = 8,
                        show_panel_grid = NULL,
                        plot_title = NULL,
                        x_axis_title = NULL,
                        y_axis_title = NULL,
                        show_title = "y",
                        show_ticks = "both",
                        show_text = "both",
                        text_italic = NULL,
                        x_text_angle = NULL,
                        y_text_angle = NULL,
                        legend_title = NULL,
                        legend_position = "right",
                        legend_direction = NULL,
                        comparisons = NULL,
                        test = "wilcox.test",
                        step_increase = 0.1,
                        map_signif_level = TRUE) {
  # Data transform
  fill <- enquo(fill)
  if (as_label(fill) == "NULL") {
    fill <- enquo(x)
  } else {
    fill <- enquo(fill)
  }
  if (flip) {
    input_data <- data %>%
      mutate(
        y = {{ x }},
        x = {{ y }},
        fill = {{ fill }}
      )
  } else {
    input_data <- data %>%
      mutate(
        x = {{ x }},
        y = {{ y }},
        fill = {{ fill }}
      )
  }
  # Plot core
  p <- ggplot(data = input_data, mapping = aes(x = x, y = y))
  if (as_label(fill) %in% colnames(input_data)) {
    p <- p + geom_boxplot(
      mapping = aes(fill = fill),
      fatten = fatten,
      width = width,
      color = "black",
      linewidth = 0.5 * 0.5 * 0.93,
      outlier.shape = NA, show.legend = show_legend
    )
  } else {
    fill <- gsub("\"", "", as_label(fill))
    p <- p + geom_boxplot(
      fill = fill,
      fatten = fatten,
      width = width,
      outlier.shape = NA,
      color = "black",
      linewidth = 0.5 * 0.5 * 0.93,
      show.legend = show_legend
    )
  }
  if (flip) {
    p <- p + scale_x_continuous(expand = c(0, 0.3))
    show_title <- "x"
    x_axis_title <- x_axis_title %||% enquo(y)
    y_axis_title <- y_axis_title %||% enquo(x)
  } else {
    p <- p + scale_y_continuous(expand = c(0, 0.3))
    show_title <- "y"
    x_axis_title <- x_axis_title %||% enquo(x)
    y_axis_title <- y_axis_title %||% enquo(y)
  }
  legend_title <- legend_title %||% enquo(fill)
  p <- p + cat_modify_plot(aspect_ratio = aspect_ratio,
    frame = frame,
    font_size = font_size,
    show_panel_grid = show_panel_grid,
    plot_title = plot_title,
    x_axis_title = x_axis_title,
    y_axis_title = y_axis_title,
    show_title = show_title,
    show_ticks = show_ticks,
    show_text = show_text,
    text_italic = text_italic,
    x_text_angle = x_text_angle,
    y_text_angle = y_text_angle,
    legend_title = legend_title,
    legend_position = legend_position,
    legend_direction = legend_direction,
    comparisons = comparisons,
    test = test,
    step_increase = step_increase,
    map_signif_level = map_signif_level
  )
  return(p)
}


R/utilities.R
#' @importFrom ggplot2 element_blank element_line element_rect element_text
#'   margin unit aes as_label enquo geom_boxplot ggplot guide_axis guide_legend
#'   guides labs scale_x_continuous scale_y_continuous quo_name geom_jitter
#'   position_jitter position_jitterdodge theme expansion waiver
#' @importFrom ggsignif geom_signif
#' @importFrom dplyr group_by summarise arrange desc pull case_when mutate
NULL

#' @importFrom magrittr %>%
##' @export
magrittr::`%>%`

#' Set a default value if an object is null
#'
#' @param lhs An object to set if it's null
#' @param rhs The value to provide if x is null
#'
#' @return rhs if lhs is null, else lhs
#'
#' @name set-if-null
#'
#' @author Hadley Wickham
#' @references \url{https://adv-r.hadley.nz/functions.html#missing-arguments}
#'
#' @examples
#' \dontrun{
#' 4 %||% 5
#' NULL %||% 5
#' }
#'
#' @keywords internal
#'
`%||%` <- function(lhs, rhs) {
  if (!is.null(x = lhs)) {
    return(lhs)
  } else {
    return(rhs)
  }
}


#' Check if a variable is a character or factor
#'
#' This function checks if a variable is a character or factor.
#'
#' @param x A variable to be checked.
#'
#' @return A logical value indicating whether the variable is a character or factor.
#'
#' @examples
#' library(catplot)
#' is_char_or_factor("hello")
#' is_char_or_factor(factor("world"))
#'
#' @export
is_char_or_factor <- function(x) {
  is.character(x) || is.factor(x)
}

#' Check if a string is a valid color
#'
#' This function attempts to convert a string to an RGB value using the
#' \code{col2rgb()} function from the \code{grDevices} package. If the RGB value
#' is not empty, the string is considered a valid color.
#'
#' @param str A character string representing a color
#' @return A logical value indicating whether the input is a valid color
#' @examples
#' library(catplot)
#' is_color("red") # TRUE
#' is_color("#00FF00") # TRUE
#' is_color("not a color") # FALSE
#' @export
is_color <- function(str) {
  # Attempt to convert the string to an RGB value
  rgb <- tryCatch(grDevices::col2rgb(str), error = function(e) NULL)

  # If the RGB value is not empty, the string is a valid color
  !is.null(rgb)
}


R/cat_boxplot.R
#' Create a box plot with customizable color or fill
#'
#' This function creates a box plot using the \code{geom_boxplot()} function
#' from the \code{ggplot2} package. The color or fill aesthetic can be
#' customized using the \code{color} or \code{fill} parameter, respectively.
#'
#' @param color A character string representing the color to use for the box
#'   plot outline
#' @param fill A character string representing the color to use for the box plot
#'   fill
#' @param ... Additional arguments to be passed to
#'   \code{\link[ggplot2]{geom_boxplot}}
#' @return A ggplot object with a box plot
#' @examples
#' library(catplot)
#' library(ggplot2)
#' # Create a box plot with red outline
#' geom_boxplot_internal(color = "red")
#'
#' # Create a box plot with green fill
#' geom_boxplot_internal(fill = "green")
#'
#' # Create a box plot with custom data
#' data <- data.frame(x = rep(c("A", "B"), each = 10), y = rnorm(20))
#' ggplot(data, aes(x, y)) +
#'   geom_boxplot_internal(color = "blue", fill = "yellow")
#' @export
geom_boxplot_internal <- function(color = NULL, fill = NULL, ...) {
  color <- enquo(color) # Capture the 'color' argument
  fill <- enquo(fill) # Capture the 'fill' argument
  if (quo_name(color) != "NULL") {
    # Check if 'color' is provided
    if (is_color(quo_name(color))) {
      # Check if 'color' is a valid color
      geom_boxplot(color = quo_name(color), ...) # Use 'color' directly
    } else {
      geom_boxplot(
        mapping = aes(
          color = !!color
        ), ...
      ) # Map 'color' using 'aes()'
    }
  } else if (quo_name(fill) != "NULL") {
    # Check if 'fill' is provided
    if (is_color(quo_name(fill))) {
      # Check if 'fill' is a valid color
      geom_boxplot(
        fill = quo_name(fill),
        color = "black", ...
      ) # Use 'fill' directly
    } else {
      geom_boxplot(
        mapping = aes(fill = !!fill),
        color = "black",
        ...
      ) # Map 'fill' using 'aes()'
    }
  } else {
    # Throw an error if neither color nor fill is provided
    stop("Please provide either 'color' or 'fill' parameter.")
  }
}

#' Create a boxplot with customized options using ggplot2
#'
#' This function creates a grouped boxplot with optional jitter points. It takes
#' in a data frame, x variable, y variable, fill variable, color variable, and
#' various other arguments to customize the plot.
#'
#' @inheritParams geom_boxplot_internal
#' @inheritParams cat_modify_data
#' @inheritParams cat_modify_plot
#' @param width A numeric value indicating the width of each boxplot. Defaults
#'   to 0.4.
#' @param fatten A numeric value indicating the amount to increase the width of
#'   each boxplot relative to the default. Defaults to 1.
#' @param outlier_colour Deprecated. Use outlier_color instead.
#' @param outlier_color A character string or symbol indicating the color of the
#'   outlier points. Defaults to NULL.
#' @param outlier_fill A character string or symbol indicating the fill color of
#'   the outlier points. Defaults to NULL.
#' @param outlier_shape An integer indicating the shape of the outlier points.
#'   Defaults to 19.
#' @param outlier_size A numeric value indicating the size of the outlier
#'   points. Defaults to 0.5.
#' @param outlier_stroke A numeric value indicating the stroke width of the
#'   outlier points. Defaults to 0.5.
#' @param outlier_alpha A numeric value indicating the alpha transparency of the
#'   outlier points. Defaults to NULL.
#' @param notch A logical value indicating whether to display notches around the
#'   median. Defaults to FALSE.
#' @param notchwidth A numeric value indicating the width of the notches.
#'   Defaults to 0.5.
#' @param varwidth A logical value indicating whether to adjust the width of
#'   each boxplot based on the number of observations. Defaults to FALSE.
#' @param na_rm A logical value indicating whether to remove missing values.
#'   Defaults to FALSE.
#' @param orientation A character string indicating whether the plot should be
#'   horizontal ("horizontal") or vertical ("vertical"). Defaults to NA.
#' @param show_legend A logical value indicating whether to show the legend.
#'   Defaults to FALSE.
#' @param add_jitter A logical value indicating whether to add jitter points to
#'   the plot. Defaults to FALSE.
#' @param jitter_width A numeric value indicating the width of the jitter
#'   points. Defaults to 0.6.
#' @param jitter_size A numeric value indicating the size of the jitter points.
#'   Defaults to 0.5.
#' @param jitter_color A character string or symbol indicating the variable to
#'   use for color of the jitter points. Defaults to NULL.
#'
#' @return A ggplot object representing the grouped boxplot with optional jitter
#'   points.
#'
#' @examples
#' library(catplot)
#' library(ggplot2)
#' library(dplyr)
#'
#' data("iris")
#' iris %>%
#'   mutate(group = if_else(log2(Petal.Length + 1) > 1.5,
#'     "high", "low"
#'   )) %>%
#'   cat_boxplot(
#'     x = Species,
#'     y = log2(Petal.Width + 1),
#'     fill = group,
#'     aspect_ratio = 1,
#'     show_panel_grid = "both",
#'     legend_title = "ee",
#'     legend_position = "top",
#'     x_text_angle = 45
#'   )
#'
#' p <- iris %>%
#'   cat_boxplot(
#'     x = Species,
#'     y = Sepal.Length,
#'     aspect_ratio = 2,
#'     frame = "open",
#'     x_text_angle = 45,
#'     comparisons = list(c("versicolor", "virginica"))
#'   ) + scale_fill_brewer(palette = "Paired")
#' p
#' @export
cat_boxplot <- function(data,
                        x,
                        y,
                        fill = NULL,
                        color = NULL,
                        x_order = NULL,
                        y_order = NULL,
                        width = 0.4,
                        fatten = 1,
                        outlier_colour = NULL,
                        outlier_color = NULL,
                        outlier_fill = NULL,
                        outlier_shape = 19,
                        outlier_size = 0.5,
                        outlier_stroke = 0.5,
                        outlier_alpha = NULL,
                        notch = FALSE,
                        notchwidth = 0.5,
                        varwidth = FALSE,
                        na_rm = FALSE,
                        orientation = NA,
                        show_legend = FALSE,
                        add_jitter = FALSE,
                        jitter_width = 0.6,
                        jitter_size = 0.5,
                        jitter_color = NULL,
                        plot_title = waiver(),
                        x_axis_title = waiver(),
                        y_axis_title = waiver(),
                        x_text_angle = NULL,
                        y_text_angle = NULL,
                        legend_title = waiver(),
                        font_size = 8,
                        linewidth = 0.5,
                        aspect_ratio = NULL,
                        frame = "closed",
                        show_panel_grid = "both_not",
                        show_title = "both",
                        show_text = "both",
                        text_italic = "both_not",
                        show_ticks = "both",
                        ticks_length = 4,
                        legend_position = "right",
                        legend_direction = NULL,
                        comparisons = NULL,
                        test = "wilcox.test",
                        step_increase = 0.1,
                        map_signif_level = TRUE) {
  # Check if x variable is character or factor
  x_is_char_or_factor <- is_char_or_factor(data[[quo_name(enquo(x))]])
  # Set color variable if provided
  if (quo_name(enquo(color)) != "NULL") {
    color <- enquo(color)
  }
  if (quo_name(enquo(fill)) != "NULL") {
    fill <- enquo(fill)
  }
  # Determine fill variable based on x and color variables
  if (quo_name(enquo(fill)) == "NULL" && quo_name(enquo(color)) == "NULL" &&
    x_is_char_or_factor) {
    fill <- enquo(x)
  } else if (!x_is_char_or_factor) {
    fill <- enquo(y)
  }
  # Set default jitter_color if not provided
  jitter_color <- enquo(jitter_color)
  if (quo_name(jitter_color) == "NULL") {
    jitter_color <- "black"
  }
  # Show legend if necessary
  if (all(
    quo_name(enquo(x)) != quo_name(fill),
    quo_name(enquo(x)) != quo_name(color)
  )) {
    show_legend <- TRUE
  }
  # Modify data based on input arguments
  data <- cat_modify_data(data = data, x = {{ x }}, y = {{ y }}, x_order = x_order, y_order = y_order)
  # Create the boxplot
  p <- ggplot(data = data, mapping = aes(x = {{ x }}, y = {{ y }})) +
    geom_boxplot_internal(
      fill = {{ fill }},
      color = {{ color }},
      width = width,
      linewidth = linewidth * 0.5 * 0.93,
      fatten = fatten,
      outlier.colour = outlier_colour,
      outlier.color = outlier_color,
      outlier.fill = outlier_fill,
      outlier.shape = outlier_shape,
      outlier.size = outlier_size,
      outlier.stroke = outlier_stroke,
      outlier.alpha = outlier_alpha,
      notch = notch,
      notchwidth = notchwidth,
      varwidth = varwidth,
      na.rm = na_rm,
      orientation = orientation,
      show.legend = show_legend
    )
  # Modify the plot based on input arguments
  p <- p + cat_modify_plot(
    plot_title = plot_title,
    x_axis_title = x_axis_title,
    y_axis_title = y_axis_title,
    x_text_angle = x_text_angle,
    y_text_angle = y_text_angle,
    legend_title = legend_title,
    font_size = font_size,
    linewidth = linewidth,
    aspect_ratio = aspect_ratio,
    frame = frame,
    show_panel_grid = show_panel_grid,
    show_title = show_title,
    show_text = show_text,
    text_italic = text_italic,
    show_ticks = show_ticks,
    ticks_length = ticks_length,
    legend_position = legend_position,
    legend_direction = legend_direction,
    comparisons = comparisons,
    test = test,
    step_increase = step_increase,
    map_signif_level = map_signif_level
  )
  # Adjust y-axis range if comparisons are provided
  if (!is.null(comparisons)) {
    scale_fn <- if (x_is_char_or_factor) scale_y_continuous else scale_x_continuous
    p <- p + scale_fn(expand = expansion(mult = c(0.05, 0.1)))
  }
  # Add jitter points if required
  if (add_jitter && quo_name(jitter_color) %in% colnames(data)) {
    p <- p + geom_jitter(
      mapping = aes(color = {{ jitter_color }}),
      size = jitter_size,
      position = position_jitterdodge(
        jitter.width = jitter_width / 2,
        jitter.height = 0.1,
        seed = 717
      ),
      show.legend = FALSE
    )
  } else if (add_jitter && !(quo_name(jitter_color) %in% colnames(data))) {
    color <- quo_name(jitter_color)
    p <- p + geom_jitter(
      color = color,
      size = jitter_size,
      position = position_jitter(
        width = jitter_width / 2,
        height = NULL, seed = 717
      ),
      show.legend = FALSE
    )
  }
  # Return the plot
  return(p)
}


R/theme.R
#' Custom theme for catplot function
#'
#' This is a custom theme for the catplot function in the catplot package. It
#' sets the visual appearance of the plot elements, such as axis, panel, legend,
#' and title, to match the default style of catplot plots.
#'
#' @param aspect_ratio numeric; ratio of x to y axis length
#' @param frame character; indicates whether the plot frame should be "open" or
#'   "closed"
#' @param show_panel_grid character; indicates which panel grid to show, "x",
#'   "y", "both", or "both_not"
#' @param show_title character; indicates which axis title to show, "x", "y",
#'   "both", or "both_not"
#' @param show_ticks character; indicates which axis ticks to show, "x", "y",
#'   "both", or "both_not"
#' @param show_text character; indicates which axis text to show, "x", "y",
#'   "both", or "both_not"
#' @param text_italic character; indicates which axis text to italicize, "x",
#'   "y", "both", or "both_not"
#' @param legend_position character; indicates the position of the legend,
#'   "right", "left", "bottom", or "top"
#' @param legend_direction character; indicates the direction of the legend, "h"
#'   or "horizontal", "v" or "vertical", or "default"
#' @param font_size numeric; base size of text elements in the theme
#' @param base_family character; base font family of text elements in the theme
#' @param base_line_size numeric; base line size of elements in the theme
#' @param base_rect_size numeric; base line size of rect in the theme
#' @param ... additional arguments passed to \code{\link[ggplot2:theme]{theme}}
#'
#' @return a ggplot2 theme object
#' @export
#'
#' @examples
#' library(catplot)
#' library(ggplot2)
#'
#' data("iris")
#' ggplot(data = iris, aes(x = Species, y = Sepal.Length)) +
#'   geom_boxplot() +
#'   theme_cat(
#'     aspect_ratio = 1,
#'     show_panel_grid = "both",
#'     show_title = "y"
#'   )
theme_cat <- function(aspect_ratio = NULL,
                      frame = "closed",
                      show_panel_grid = NULL,
                      show_title = "both",
                      show_ticks = "both",
                      show_text = "both",
                      text_italic = NULL,
                      legend_position = "right",
                      legend_direction = NULL,
                      font_size = 8,
                      base_family = "",
                      base_line_size = font_size / 16,
                      base_rect_size = font_size / 16,
                      ...) {
  match.arg(arg = frame, choices = c("closed", "open"))
  show_panel_grid <- show_panel_grid %||% "both_not"
  match.arg(arg = show_panel_grid,
            choices = c("x", "y", "both", "both_not"))
  show_title <- show_title %||% "both_not"
  match.arg(arg = show_title,
            choices = c("x", "y", "both", "both_not"))
  show_ticks <- show_ticks %||% "both_not"
  match.arg(arg = show_ticks,
            choices = c("x", "y", "both", "both_not"))
  show_text <- show_text %||% "both_not"
  match.arg(arg = show_text,
            choices = c("x", "y", "both", "both_not"))
  text_italic <- text_italic %||% "both_not"
  match.arg(arg = text_italic,
            choices = c("x", "y", "both", "both_not"))
  match.arg(arg = legend_position,
            choices = c("right", "left", "bottom", "top"))

  legend_margin <- switch(
    legend_position,
    "right" = margin(l = -8),
    "top" = margin(b = -8),
    "bottom" = margin(t = -8),
    "left" = margin(r = -8)
  )
  legend_direction <- legend_direction %||% "default"
  legend_direction <-
    if (legend_direction %in% c("h", "horizontal")) {
      "horizontal"
    } else if (legend_direction %in% c("v", "vertical")) {
      "vertical"
    } else if (legend_direction == "default") {
      legend_direction <- NULL
    }
  theme <- theme(
    aspect.ratio = aspect_ratio,
    # Axis
    axis.ticks = element_blank(),
    axis.title = element_blank(),
    axis.text = element_blank(),
    axis.ticks.length = unit(4, "pt"),
    # Panel
    panel.background = element_blank(),
    panel.grid = element_blank(),
    # Legend
    legend.background = element_blank(),
    legend.title = element_text(
      size = font_size,
      face = "plain",
      colour = "black"
    ),
    legend.text = element_text(
      size = font_size,
      face = "plain",
      colour = "black"
    ),
    legend.key = element_blank(),
    legend.key.height = unit(font_size, "pt"),
    legend.key.width = unit(font_size, "pt"),
    legend.position = legend_position,
    legend.margin = legend_margin,
    legend.direction = legend_direction,
    # Plot
    plot.background = element_blank(),
    plot.title = element_text(
      size = font_size,
      hjust = 0.5,
      vjust = -1,
      face = "plain",
      colour = "black"
    ),
    plot.subtitle = element_text(
      hjust = 0.5,
      size = font_size,
      face = "plain",
      colour = "black"
    ),
    # Facetting
    strip.background = element_blank(),
    strip.text = element_text(
      size = font_size,
      face = "plain",
      colour = "black"
    ),
    ...
  )
  # Frame
  if (frame == "open") {
    theme <- theme + theme(
      axis.line = element_line(
        linewidth = base_line_size * 0.5 / 1.07,
        lineend = "square",
        colour = "black"
      ),
      panel.border = element_blank()
    )
  } else if (frame == "closed") {
    theme <- theme + theme(
      axis.line = element_blank(),
      panel.border = element_rect(
        linewidth = base_line_size * 0.5 / 1.07,
        fill = NA,
        colour = "black"
      )
    )
  }
  # Title
  if (show_title == "x") {
    theme <- theme + theme(
      axis.title.x = element_text(
        size = font_size,
        face = "plain",
        colour = "black"
      ),
      axis.title.y = element_blank()
    )
  } else if (show_title == "y") {
    theme <- theme + theme(
      axis.title.y = element_text(
        size = font_size,
        face = "plain",
        colour = "black"
      ),
      axis.title.x = element_blank()
    )
  } else if (show_title == "both") {
    theme <- theme + theme(axis.title = element_text(
      size = font_size,
      face = "plain",
      colour = "black"
    ))
  }
  # Ticks
  if (show_ticks == "x") {
    theme <- theme +
      theme(
        axis.ticks.x = element_line(
          linewidth = base_line_size * 0.5 / 1.07,
          lineend = "square",
          colour = "black"
        ),
        axis.ticks.y = element_blank()
      )
  } else if (show_ticks == "y") {
    theme <- theme +
      theme(
        axis.ticks.y = element_line(
          linewidth = base_line_size * 0.5 / 1.07,
          lineend = "square",
          colour = "black"
        ),
        axis.ticks.x = element_blank()
      )
  } else if (show_ticks == "both") {
    theme <- theme +
      theme(
        axis.ticks = element_line(
          linewidth = base_line_size * 0.5 / 1.07,
          lineend = "square",
          colour = "black"
        )
      )
  }
  # Fontface
  x_fontface <- ifelse(text_italic == "x" |
                         text_italic == "both", "italic", "plain")
  y_fontface <- ifelse(text_italic == "y" |
                         text_italic == "both", "italic", "plain")
  # Text
  if (show_text == "x") {
    theme <- theme +
      theme(
        axis.text.x = element_text(
          size = font_size,
          face = x_fontface,
          colour = "black"
        ),
        axis.text.y = element_blank()
      )
  } else if (show_text == "y") {
    theme <- theme +
      theme(
        axis.text.y = element_text(
          size = font_size,
          face = y_fontface,
          colour = "black"
        ),
        axis.text.x = element_blank()
      )
  } else if (show_text == "both") {
    theme <- theme +
      theme(
        axis.text.x = element_text(
          size = font_size,
          face = x_fontface,
          colour = "black"
        ),
        axis.text.y = element_text(
          size = font_size,
          face = x_fontface,
          colour = "black"
        )
      )
  }
  # Panel grid
  if (show_panel_grid == "x") {
    theme <- theme +
      theme(
        panel.grid.major.x = element_line(
          colour = "lightgrey",
          linewidth = base_line_size * 0.5 / 1.07,
          lineend = "square"
        ),
        panel.grid.major.y = element_blank()
      )
  } else if (show_panel_grid == "y") {
    theme <- theme +
      theme(
        panel.grid.major.y = element_line(
          colour = "lightgrey",
          linewidth = base_line_size * 0.5 / 1.07,
          lineend = "square"
        ),
        panel.grid.major.x = element_blank()
      )
  } else if (show_panel_grid == "both") {
    theme <- theme +
      theme(
        panel.grid = element_line(
          colour = "lightgrey",
          linewidth = base_line_size * 0.5 / 1.07,
          lineend = "square"
        )
      )
  }
  theme <- theme + theme(...)
  return(theme)
}


#' Modify a ggplot2 plot with customized options and add significance labels
#'
#' @inheritParams theme_cat
#' @param plot_title Title of the plot.
#' @param x_axis_title Title of the x-axis.
#' @param y_axis_title Title of the y-axis.
#' @param x_text_angle Angle of the x-axis labels.
#' @param y_text_angle Angle of the y-axis labels.
#' @param legend_title Title of the legend.
#' @param legend_position Position of the legend, can be "left", "right", "top", "bottom", or NULL.
#' @param legend_direction Direction of the legend, can be "horizontal" or "vertical".
#' @param comparisons List of comparison groups for adding significance labels.
#' @param test Type of statistical test to use for calculating significance. Defaults to "wilcox.test".
#' @param step_increase Step increase for adjusting the significance label position.
#' @param map_signif_level Logical indicating whether to map the significance level to asterisks.
#'
#' @return A list of ggplot2 options for modifying a plot and adding significance labels.
#'
#' @examples
#' library(ggplot2)
#'
#' p <- ggplot(mtcars, aes(x = factor(cyl), y = mpg)) +
#'   geom_boxplot() +
#'   cat_modify_plot(plot_title = "My Plot Title",
#'                 x_axis_title = "X Axis Title",
#'                 y_axis_title = "Y Axis Title",
#'                 test = "t.test",
#'                 comparisons = list(c("4", "6"), c("4", "8")),
#'                 aspect_ratio = 1)
#' p
#'
#' @export
cat_modify_plot <- function(plot_title = NULL,
                            x_axis_title = NULL,
                            y_axis_title = NULL,
                            x_text_angle = NULL,
                            y_text_angle = NULL,
                            legend_title = NULL,
                            comparisons = NULL,
                            test = "wilcox.test",
                            step_increase = 0.1,
                            map_signif_level = TRUE,
                            aspect_ratio = NULL,
                            frame = "closed",
                            font_size = 8,
                            show_panel_grid = NULL,
                            show_title = "both",
                            show_ticks = "both",
                            show_text = "both",
                            text_italic = NULL,
                            legend_position = "right",
                            legend_direction = NULL) {
  list(
    # Plot title and axis title
    labs(title = plot_title, x = x_axis_title, y = y_axis_title),
    # Plot legend
    guides(
      x = guide_axis(angle = x_text_angle),
      y = guide_axis(angle = y_text_angle),
      fill = guide_legend(title = legend_title)
    ),
    # Plot ggsignif
    if (!is.null(comparisons)) {
      geom_signif(
        comparisons = comparisons,
        size = 0.5 * 0.5 * 0.93,
        textsize = 0.36 * 6,
        test = test,
        step_increase = step_increase,
        tip_length = 0.02,
        map_signif_level = map_signif_level
      )
    },
    # Plot theme
    theme_cat(
      aspect_ratio = aspect_ratio,
      frame = frame,
      show_panel_grid = show_panel_grid,
      show_title = show_title,
      show_ticks = show_ticks,
      show_text = show_text,
      text_italic = text_italic,
      legend_position = legend_position,
      legend_direction = legend_direction,
      font_size = font_size
    )
  )
}

# cat_modify_plot <- function(p,
#                             plot_title = NULL,
#                             x_axis_title = NULL,
#                             y_axis_title = NULL,
#                             comparisons = NULL,
#                             test = "wilcox.test",
#                             step_increase = 0.1,
#                             map_signif_level = TRUE,
#                             flip = FALSE,
#                             ) {
#   # Plot title and axis title
#   p <-
#     p + labs(title = plot_title, x = x_axis_title, y = y_axis_title)
#   # Plot theme
#   p <- p +
#     theme_cat(
#       aspect_ratio = aspect_ratio,
#       frame = frame,
#       show_panel_grid = show_panel_grid,
#       show_title = show_title,
#       show_ticks = show_ticks,
#       show_text = show_text,
#       text_italic = text_italic,
#       legend_position = legend_position,
#       legend_direction = legend_direction,
#       font_size = font_size
#     )
#   # Plot legend
#   p <- p + guides(
#     x = guide_axis(angle = x_text_angle),
#     y = guide_axis(angle = y_text_angle),
#     fill = guide_legend(title = legend_title)
#   )
#
#   # Plot scale
#   if (flip) {
#     p <- p +
#       scale_x_continuous(expand = c(0, 0.3))
#   } else {
#     p <- p +
#       scale_y_continuous(expand = c(0, 0.3))
#   }
#
#   # Plot ggsignif
#   if (!is.null(comparisons)) {
#     p <- p + geom_signif(
#       comparisons = comparisons,
#       size = 0.5 * 0.5 * 0.93,
#       textsize = 0.36 * 6,
#       test = test,
#       step_increase = step_increase,
#       tip_length = 0.02,
#       map_signif_level = map_signif_level
#     )
#   }
#   return(p)
# }
